### DISCUSSION TITLE
How can I get the target URI of a standalone ref:`foo` reference?

### DISCUSSION THREAD

**Discussion by lelit** [sphinx-doc/sphinx#12792] [Q&A] [answered]
I'm writing a [Sphinx extension](https://gitlab.com/metapensiero/metapensiero.sphinx.d2) that implements a directive able to render a [d2 diagram](https://d2lang.com/) and insert the resulting `PNG` or `SVG` into a document, very similar to what [ext/graphviz.py](https://github.com/sphinx-doc/sphinx/blob/master/sphinx/ext/graphviz.py) does.

It is already good enough for my needs, but I'd like to exploit the [link](https://d2lang.com/tour/interactive/#links) feature of a `d2` node so that I can use `:ref:` to set the target to an arbitrary Sphinx reference.

In other words, instead of writing this:

```
.. d2::
   Organization: {
     shape: class
     link: /uri/pointing/to/the/python/Organization/class.html
     name: str
   }
```

I would prefer being able to say:

```
.. d2::
   Organization: {
     shape: class
     link: :py:class:`Organization`
     name: str
   }
```

Given that the actual rendering of the diagram (that is, the execution of `d2` to generate the `SVG`) is already happening when the whole documentation has been built, I think it should be a matter of manually _pre-processing_ the diagram's code searching for those `link: XXX` _options_ and, when `XXX` is a reST role, process it in the current _context_ to compute the final URI and replace `XXX` with that.

I spent a few hours searching and trying, but I could not figure out a way to do that at [render time](https://gitlab.com/metapensiero/metapensiero.sphinx.d2/-/blob/71efbcfa25aa032fe9eb6cd5eaca3a8a436a5443/src/metapensiero/sphinx/d2/__init__.py#L158).

Can you shed some light on how I could do that?

Thanks in advance for any hint!

**Comment by AA-Turner** [comment]
You will need to do this in the reading phase (in your directive). Once you've extracted the reST to process, call [`self.parse_inline`](https://www.sphinx-doc.org/en/master/extdev/utils.html#sphinx.util.docutils.SphinxDirective.parse_inline) and use the URI of the result node.

A

**Reply by lelit** [reply]
Thanks, I tried that, but doing it in the reading phase, I failed to see how I can _resolve_ that to the final target URI, because at that time the _location_ of the Python class could be not known yet... what am I missing?

**Reply by AA-Turner** [reply]
Oh, if you're getting a `pending_xref` then I would store that somewhere (perhaps as a child of your custom node) and then let the ReferencesResolver handle them.

A

**Reply by lelit** [reply, marked as answer]
Yay, I think I got it!

Basically
- in the reading phase, I collect those links, parse them and add resulting nodes to the `d2node` produced by the directive.
- later, when rendering the diagram, I first replace all those links with _subnodes_ `refuri`/`refid`

Thanks a lot!

**Reply by lelit** [reply]
This [commit](https://gitlab.com/metapensiero/metapensiero.sphinx.d2/-/commit/b0dcf075e4f38ffff0da1af4d5e38817e1ad334f) implements the desired facility, but I need to study further to understand how to treat a local reference (that is, when the reference has just a `refid`) [here](https://gitlab.com/metapensiero/metapensiero.sphinx.d2/-/blob/b0dcf075e4f38ffff0da1af4d5e38817e1ad334f/src/metapensiero/sphinx/d2/__init__.py#L189).

The case is [this one](https://gitlab.com/metapensiero/metapensiero.sphinx.d2/-/blob/b0dcf075e4f38ffff0da1af4d5e38817e1ad334f/example/index.rst?plain=1#L26), where the reference is within the very same document containing the diagram: when I am about replacing the original ":ref:`d2`", the resolved reference contains a `refid` set to `d2`. I need to substitute the original with something like `/index.html#d2`, not just `#d2` because the latter one will be relative to the embedded `SVG` object, and obviously wrong. I failed to find a way to determine that `index.html`, that is the base URI of the containing document page. Can you point me in the right direction?

Thanks again!

**Reply by lelit** [reply]
Later I will try to see whether the [fix_svg_relative_paths() function](https://github.com/sphinx-doc/sphinx/blob/master/sphinx/ext/graphviz.py#L224) does right that...

**Reply by lelit** [reply]
For the record, no, that is unrelated, but it gave me an hint about using `env.path2doc(self.document["source"])`.
All is good (for now): https://pypi.org/project/metapensiero.sphinx.d2/

**Reply by AA-Turner** [reply]
Fab, do let me know if there are improvements we can make here in Sphinx.

A

**Reply by lelit** [reply]
One thing comes to mind: how about having an _enhanced_ version of `docutils.directives.flag` similar to [this one](https://gitlab.com/metapensiero/metapensiero.sphinx.d2/-/blob/v0.1/src/metapensiero/sphinx/d2/__init__.py?ref_type=tags#L54-64), that would allow to _negate_ a global setting, perhaps introduced by a configuration value, as I did [for example](https://gitlab.com/metapensiero/metapensiero.sphinx.d2/-/blob/v0.1/src/metapensiero/sphinx/d2/__init__.py?ref_type=tags#L333-335) for the `:center:` flag?

**Reply by lelit** [reply]
I had to do some more [gymnic](https://gitlab.com/metapensiero/metapensiero.sphinx.d2/-/blob/v0.3/src/metapensiero/sphinx/d2/__init__.py?ref_type=tags#L191-208) to absolutize resolved refs URIs, I could not find an helper function: if I didn't miss it, maybe the `builder` could expose something like that?

**Comment by AA-Turner** [comment]
Entirely off topic, but out of interest are you the same Lele that contributed Italian translations etc to Docutils (e.g. at [r2533](https://sourceforge.net/p/docutils/code/2533))? We're (slowly) working on a migration to Git, so matching people up is useful!

Best,
Adam

**Reply by lelit** [reply]
Yes I am!
